Wrong operation name for routes wrapped with flatMaped directives

I have encountered this issue after migration to Scala 2.13 and Kamon 2.0. The simplest way to reproduce it is to add authenticateBasic wrapper in the route of TestWebServer.
```
      pathPrefix("extraction") {
        authenticateBasic("realm", credentials => Option("Okay")) { srt =>
          (post | get) {
            pathPrefix("nested") {
              pathPrefix(IntNumber / "fixed") { num =>
                pathPrefix("anchor" / IntNumber.? / JavaUUID / "fixed") { (number, uuid) =>
                  pathPrefix(LongNumber / HexIntNumber) { (longNum, hex) =>
                    complete("OK")
                  }
                }
              }
            } ~
            pathPrefix("concat") {
              path("fixed" ~ JavaUUID ~ HexIntNumber) { (uuid, num) =>
                complete("OK")
              }
            } ~
            pathPrefix("on-complete" / IntNumber) { _ =>
              onComplete(Future("hello")) { _ =>
                extract(samplingDecision) { decision =>
                  path("more-path") {
                    complete(decision.toString)
                  }
                }
              }
            } ~
            pathPrefix("on-success" / IntNumber) { _ =>
              onSuccess(Future("hello")) { text =>
                pathPrefix("after") {
                  complete(text)
                }
              }
            } ~
            pathPrefix("complete-or-recover-with" / IntNumber) { _ =>
              completeOrRecoverWith(Future("bad".charAt(10).toString)) { failure =>
                pathPrefix("after") {
                  failWith(failure)
                }
              }
            } ~
            pathPrefix("complete-or-recover-with-success" / IntNumber) { _ =>
              completeOrRecoverWith(Future("good")) { failure =>
                pathPrefix("after") {
                  failWith(failure)
                }
              }
            }
          }
        }
      } 
```
Test results look like this
```
[info] FastFutureInstrumentationSpec:
[info] the FastFuture instrumentation
[info]   should keep the Context captured by the Future from which it was created
[info]   - when calling .map/.flatMap/.onComplete and the original Future has not completed yet
[info]   - when calling .map/.flatMap/.onComplete and the original Future has already completed
[info] AkkaHttpServerTracingSpec:
[info] the Akka HTTP server instrumentation
[info] - should create a server Span when receiving requests
[info]   should not include variables in operation name
[info]   - when including nested directives *** FAILED ***
[info]     The code passed to eventually never returned normally. Attempted 567 times over 10.007980149000002 seconds. Last failure message: The Option on which value was invoked was not defined.. (AkkaHttpServerTracingSpec.scala:68)
[info]   - when take a sampling decision when the routing tree hits an onComplete directive *** FAILED ***
[info]     The code passed to eventually never returned normally. Attempted 573 times over 10.016707381999998 seconds. Last failure message: The Option on which value was invoked was not defined.. (AkkaHttpServerTracingSpec.scala:80)
[info]   - when take a sampling decision when the routing tree hits an onSuccess directive *** FAILED ***
[info]     The code passed to eventually never returned normally. Attempted 579 times over 10.009030289999998 seconds. Last failure message: The Option on which value was invoked was not defined.. (AkkaHttpServerTracingSpec.scala:92)
[ERROR] [07/29/2019 18:02:46.191] [http-server-instrumentation-spec-akka.actor.default-dispatcher-2] [akka.actor.ActorSystemImpl(http-server-instrumentation-spec)] Error during processing of request: 'String index out of range: 10'. Completing with 500 Internal Server Error response. To change default exception handling behavior, provide a custom ExceptionHandler.
java.lang.StringIndexOutOfBoundsException: String index out of range: 10
	at java.lang.String.charAt(String.java:658)
	at kamon.testkit.TestWebServer.$anonfun$startServer$33(TestWebServer.scala:92)
	at scala.concurrent.Future$.$anonfun$apply$1(Future.scala:660)
	at scala.concurrent.impl.Promise$Transformation.run(Promise.scala:430)
	at akka.dispatch.BatchingExecutor$AbstractBatch.processBatch(BatchingExecutor.scala:55)
	at akka.dispatch.BatchingExecutor$BlockableBatch.$anonfun$run$1(BatchingExecutor.scala:92)
	at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.scala:18)
	at scala.concurrent.BlockContext$.withBlockContext(BlockContext.scala:94)
	at akka.dispatch.BatchingExecutor$BlockableBatch.run(BatchingExecutor.scala:92)
	at akka.dispatch.TaskInvocation.run(AbstractDispatcher.scala:41)
	at kamon.instrumentation.executor.ExecutorInstrumentation$InstrumentedForkJoinPool$TimingRunnable.run(ExecutorInstrumentation.scala:653)
	at akka.dispatch.ForkJoinExecutorConfigurator$AkkaForkJoinTask.exec(ForkJoinExecutorConfigurator.scala:49)
	at akka.dispatch.forkjoin.ForkJoinTask.doExec(ForkJoinTask.java:260)
	at akka.dispatch.forkjoin.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1339)
	at akka.dispatch.forkjoin.ForkJoinPool.runWorker(ForkJoinPool.java:1979)
	at akka.dispatch.forkjoin.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:107)

[info]   - when take a sampling decision when the routing tree hits a completeOrRecoverWith directive with a failed future *** FAILED ***
[info]     The code passed to eventually never returned normally. Attempted 590 times over 10.008884834 seconds. Last failure message: The Option on which value was invoked was not defined.. (AkkaHttpServerTracingSpec.scala:104)
[info]   - when take a sampling decision when the routing tree hits a completeOrRecoverWith directive with a successful future *** FAILED ***
[info]     The code passed to eventually never returned normally. Attempted 584 times over 10.01066485 seconds. Last failure message: The Option on which value was invoked was not defined.. (AkkaHttpServerTracingSpec.scala:116)
[info]   - when including ambiguous nested directives
[info] - should change the Span operation name when using the operationName directive
[info] - should mark spans as failed when request fails
[info] - should change the operation name to 'unhandled' when the response status code is 404
[info] - should correctly time entity transfer timings
[info] - should include the trace-id and keep all user-provided headers in the responses
[info] AkkaHttpServerMetricsSpec:
[info] the Akka HTTP server instrumentation
[info] - should track the number of open connections and active requests on the Server side
[info] AkkaHttpClientTracingSpec:
[info] the Akka HTTP client instrumentation
[info] - should create a client Span when using the request level API - Http().singleRequest(...)
[info] - should serialize the current context into HTTP Headers
[info] - should mark Spans as errors if the client request failed
[info] Run completed in 57 seconds, 326 milliseconds.
[info] Total number of tests run: 18
[info] Suites: completed 4, aborted 0
[info] Tests: succeeded 13, failed 5, canceled 0, ignored 0, pending 0
[info] *** 5 TESTS FAILED ***
```

--- a/kamon-akka-http/src/main/scala-2.11/kamon/instrumentation/akka/http/AkkaHttpServerInstrumentation.scala
+++ b/kamon-akka-http/src/main/scala-2.11/kamon/instrumentation/akka/http/AkkaHttpServerInstrumentation.scala
@@ -23,6 +23,8 @@ import scala.util.control.NonFatal
 import scala.util.{Failure, Success, Try}
 import java.util.regex.Pattern
 
+import kamon.context.Context
+
 
 class AkkaHttpServerInstrumentation extends InstrumentationBuilder {
 
@@ -159,22 +161,25 @@ object ResolveOperationNameOnRouteInterceptor {
     }
 
   private def resolveOperationName(requestContext: RequestContext): RequestContext = {
-    val defaultOperationName = ServerFlowWrapper.defaultOperationName(requestContext.request.uri.authority.port)
 
-    // We will only change the operation name if no change was applied to it. At this point, the only way in which it
-    // might have changed is if the user changed it with the operationName directive or just accessing the Span and
-    // changing it there, so we wouldn't want to overwrite that.
-    //
-    if(Kamon.currentSpan().operationName() == defaultOperationName) {
-      val allMatches = requestContext.asInstanceOf[HasMatchingContext].matchingContext.reverse.map(singleMatch)
-      val operationName = allMatches.mkString("")
+    // We will only change the operation name if the last edit made to it was an automatic one. At this point, the only
+    // way in which the operation name might have changed is if the user changed it with the operationName directive or
+    // by accessing the Span and changing it directly there, so we wouldn't want to overwrite that.
 
-      if(operationName.nonEmpty) {
-        Kamon.currentSpan()
-          .name(operationName)
-          .takeSamplingDecision()
+    Kamon.currentContext().get(LastAutomaticOperationNameEdit.Key).foreach(lastEdit => {
+      if(Kamon.currentSpan().operationName() == lastEdit.operationName) {
+        val allMatches = requestContext.asInstanceOf[HasMatchingContext].matchingContext.reverse.map(singleMatch)
+        val operationName = allMatches.mkString("")
+
+        if(operationName.nonEmpty) {
+          Kamon.currentSpan()
+            .name(operationName)
+            .takeSamplingDecision()
+
+          lastEdit.operationName = operationName
+        }
       }
-    }
+    })
 
     requestContext
   }
@@ -201,6 +206,16 @@ object ResolveOperationNameOnRouteInterceptor {
         }
     }
   }
+}
+
+
+class LastAutomaticOperationNameEdit(@volatile var operationName: String)
+
+object LastAutomaticOperationNameEdit {
+  val Key = Context.key[Option[LastAutomaticOperationNameEdit]]("laone", None)
+
+  def apply(operationName: String): LastAutomaticOperationNameEdit =
+    new LastAutomaticOperationNameEdit(operationName)
 }
 
 object RequestContextCopyInterceptor {

--- a/kamon-akka-http/src/main/scala-2.12/kamon/instrumentation/akka/http/AkkaHttpServerInstrumentation.scala
+++ b/kamon-akka-http/src/main/scala-2.12/kamon/instrumentation/akka/http/AkkaHttpServerInstrumentation.scala
@@ -23,6 +23,8 @@ import scala.util.control.NonFatal
 import scala.util.{Failure, Success, Try}
 import java.util.regex.Pattern
 
+import kamon.context.Context
+
 
 class AkkaHttpServerInstrumentation extends InstrumentationBuilder {
 
@@ -157,22 +159,25 @@ object ResolveOperationNameOnRouteInterceptor {
     }
 
   private def resolveOperationName(requestContext: RequestContext): RequestContext = {
-    val defaultOperationName = ServerFlowWrapper.defaultOperationName(requestContext.request.uri.authority.port)
 
-    // We will only change the operation name if no change was applied to it. At this point, the only way in which it
-    // might have changed is if the user changed it with the operationName directive or just accessing the Span and
-    // changing it there, so we wouldn't want to overwrite that.
-    //
-    if(Kamon.currentSpan().operationName() == defaultOperationName) {
-      val allMatches = requestContext.asInstanceOf[HasMatchingContext].matchingContext.reverse.map(singleMatch)
-      val operationName = allMatches.mkString("")
+    // We will only change the operation name if the last edit made to it was an automatic one. At this point, the only
+    // way in which the operation name might have changed is if the user changed it with the operationName directive or
+    // by accessing the Span and changing it directly there, so we wouldn't want to overwrite that.
 
-      if(operationName.nonEmpty) {
-        Kamon.currentSpan()
-          .name(operationName)
-          .takeSamplingDecision()
+    Kamon.currentContext().get(LastAutomaticOperationNameEdit.Key).foreach(lastEdit => {
+      if(Kamon.currentSpan().operationName() == lastEdit.operationName) {
+        val allMatches = requestContext.asInstanceOf[HasMatchingContext].matchingContext.reverse.map(singleMatch)
+        val operationName = allMatches.mkString("")
+
+        if(operationName.nonEmpty) {
+          Kamon.currentSpan()
+            .name(operationName)
+            .takeSamplingDecision()
+
+          lastEdit.operationName = operationName
+        }
       }
-    }
+    })
 
     requestContext
   }
@@ -199,6 +204,15 @@ object ResolveOperationNameOnRouteInterceptor {
         }
     }
   }
+}
+
+class LastAutomaticOperationNameEdit(@volatile var operationName: String)
+
+object LastAutomaticOperationNameEdit {
+  val Key = Context.key[Option[LastAutomaticOperationNameEdit]]("laone", None)
+
+  def apply(operationName: String): LastAutomaticOperationNameEdit =
+    new LastAutomaticOperationNameEdit(operationName)
 }
 
 object RequestContextCopyInterceptor {

--- a/kamon-akka-http/src/main/scala-2.13/kamon/instrumentation/akka/http/AkkaHttpServerInstrumentation.scala
+++ b/kamon-akka-http/src/main/scala-2.13/kamon/instrumentation/akka/http/AkkaHttpServerInstrumentation.scala
@@ -18,10 +18,12 @@ import kanela.agent.api.instrumentation.InstrumentationBuilder
 import kanela.agent.api.instrumentation.mixin.Initializer
 import kanela.agent.libs.net.bytebuddy.implementation.bind.annotation._
 
-import scala.concurrent.{ExecutionContext, Future, Promise, Batchable}
+import scala.concurrent.{Batchable, ExecutionContext, Future, Promise}
 import scala.util.control.NonFatal
 import scala.util.{Failure, Success, Try}
 import java.util.regex.Pattern
+
+import kamon.context.Context
 
 
 class AkkaHttpServerInstrumentation extends InstrumentationBuilder {
@@ -157,22 +159,25 @@ object ResolveOperationNameOnRouteInterceptor {
     }
 
   private def resolveOperationName(requestContext: RequestContext): RequestContext = {
-    val defaultOperationName = ServerFlowWrapper.defaultOperationName(requestContext.request.uri.authority.port)
 
-    // We will only change the operation name if no change was applied to it. At this point, the only way in which it
-    // might have changed is if the user changed it with the operationName directive or just accessing the Span and
-    // changing it there, so we wouldn't want to overwrite that.
-    //
-    if(Kamon.currentSpan().operationName() == defaultOperationName) {
-      val allMatches = requestContext.asInstanceOf[HasMatchingContext].matchingContext.reverse.map(singleMatch)
-      val operationName = allMatches.mkString("")
+    // We will only change the operation name if the last edit made to it was an automatic one. At this point, the only
+    // way in which the operation name might have changed is if the user changed it with the operationName directive or
+    // by accessing the Span and changing it directly there, so we wouldn't want to overwrite that.
 
-      if(operationName.nonEmpty) {
-        Kamon.currentSpan()
-          .name(operationName)
-          .takeSamplingDecision()
+    Kamon.currentContext().get(LastAutomaticOperationNameEdit.Key).foreach(lastEdit => {
+      if(Kamon.currentSpan().operationName() == lastEdit.operationName) {
+        val allMatches = requestContext.asInstanceOf[HasMatchingContext].matchingContext.reverse.map(singleMatch)
+        val operationName = allMatches.mkString("")
+
+        if(operationName.nonEmpty) {
+          Kamon.currentSpan()
+            .name(operationName)
+            .takeSamplingDecision()
+
+          lastEdit.operationName = operationName
+        }
       }
-    }
+    })
 
     requestContext
   }
@@ -199,6 +204,15 @@ object ResolveOperationNameOnRouteInterceptor {
         }
     }
   }
+}
+
+class LastAutomaticOperationNameEdit(@volatile var operationName: String)
+
+object LastAutomaticOperationNameEdit {
+  val Key = Context.key[Option[LastAutomaticOperationNameEdit]]("laone", None)
+
+  def apply(operationName: String): LastAutomaticOperationNameEdit =
+    new LastAutomaticOperationNameEdit(operationName)
 }
 
 object RequestContextCopyInterceptor {

--- a/kamon-akka-http/src/main/scala/kamon/instrumentation/akka/http/ServerFlowWrapper.scala
+++ b/kamon-akka-http/src/main/scala/kamon/instrumentation/akka/http/ServerFlowWrapper.scala
@@ -69,7 +69,10 @@ object ServerFlowWrapper {
 
           // The only reason why it's safe to leave the Thread dirty is because the Actor
           // instrumentation will cleanup afterwards.
-          Kamon.storeContext(requestHandler.context)
+          Kamon.storeContext(requestHandler.context.withEntry(
+            LastAutomaticOperationNameEdit.Key, Option(LastAutomaticOperationNameEdit(requestHandler.span.operationName()))
+          ))
+
           push(requestOut, request)
         }
 

--- a/kamon-akka-http/src/test/scala/kamon/akka/http/AkkaHttpServerTracingSpec.scala
+++ b/kamon-akka-http/src/test/scala/kamon/akka/http/AkkaHttpServerTracingSpec.scala
@@ -90,7 +90,7 @@ class AkkaHttpServerTracingSpec extends WordSpecLike with Matchers with ScalaFut
 
       "take a sampling decision when the routing tree hits an onComplete directive" in {
         val path = "extraction/on-complete/42/more-path"
-        val expected = "/extraction/on-complete/{}"
+        val expected = "/extraction/on-complete/{}/more-path"
         val target = s"http://$interface:$port/$path"
         Http().singleRequest(HttpRequest(uri = target)).map(_.discardEntityBytes())
 
@@ -102,7 +102,7 @@ class AkkaHttpServerTracingSpec extends WordSpecLike with Matchers with ScalaFut
 
       "take a sampling decision when the routing tree hits an onSuccess directive" in {
         val path = "extraction/on-success/42/after"
-        val expected = "/extraction/on-success/{}"
+        val expected = "/extraction/on-success/{}/after"
         val target = s"http://$interface:$port/$path"
         Http().singleRequest(HttpRequest(uri = target)).map(_.discardEntityBytes())
 
@@ -114,7 +114,7 @@ class AkkaHttpServerTracingSpec extends WordSpecLike with Matchers with ScalaFut
 
       "take a sampling decision when the routing tree hits a completeOrRecoverWith directive with a failed future" in {
         val path = "extraction/complete-or-recover-with/42/after"
-        val expected = "/extraction/complete-or-recover-with/{}"
+        val expected = "/extraction/complete-or-recover-with/{}/after"
         val target = s"http://$interface:$port/$path"
         Http().singleRequest(HttpRequest(uri = target)).map(_.discardEntityBytes())
 

--- a/kamon-akka-http/src/test/scala/kamon/testkit/TestWebServer.scala
+++ b/kamon-akka-http/src/test/scala/kamon/testkit/TestWebServer.scala
@@ -56,6 +56,7 @@ trait TestWebServer extends TracingDirectives {
           complete("OK")
         } ~
         pathPrefix("extraction") {
+          authenticateBasic("realm", credentials => Option("Okay")) { srt =>
           (post | get) {
             pathPrefix("nested") {
               pathPrefix(IntNumber / "fixed") { num =>
@@ -104,6 +105,7 @@ trait TestWebServer extends TracingDirectives {
             path("segment" / Segment){ segment =>
               complete(HttpResponse(entity = HttpEntity(ContentTypes.`text/plain(UTF-8)`, segment)))
             }
+          }
           }
         } ~
         path(rootOk) {

